kimchi/src/circuits/constraints.rs

```diff
@@ -682,12 +700,33 @@ impl<F: PrimeField + SquareRootField> Builder<F> {
 
         //~ 2. Create a domain for the circuit. That is,
         //~    compute the smallest subgroup of the field that
-        //~    has order greater or equal to `n + ZK_ROWS` elements.
-        let domain_size_lower_bound =
-            std::cmp::max(gates.len(), num_lookups + 1) + ZK_ROWS as usize;
+        //~    has order greater or equal to `n + zk_rows` elements.
+        let (zk_rows, domain_size_lower_bound) = {
+            let mut zk_rows = 3;
+            let get_domain_size_lower_bound =
+                |zk_rows: u64| std::cmp::max(gates.len(), num_lookups + 1) + zk_rows as usize;
+            let mut domain_size_lower_bound = get_domain_size_lower_bound(zk_rows);
+            if let Some(max_poly_size) = self.max_poly_size {
+                // Iterate to find a fixed-point where zk_rows is sufficient for the number of
+                // chunks that we use, and also does not cause us to overflow the domain size.
+                // NB: We use iteration here rather than hard-coding an assumption about
+                // `compute_size_of_domain`s internals. In practice, this will never be executed
+                // more than once.
+                while {
+                    let domain_size = D::<F>::compute_size_of_domain(domain_size_lower_bound)
+                        .ok_or(SetupError::DomainCreation(
+                            "could not compute size of domain",
+                        ))?;
+                    zk_rows = ((16 * (domain_size / max_poly_size) + 4) / 7) as u64;
```

Comment:
The `7` constant seems magic to my eyes. Could we give it a name? Or a comment ;-)